### PR TITLE
Harden XML parsing via commons-secure-xml

### PR DESCRIPTION
**Pull request by ppkarwasz**
Creates `XmlStringLookup`'s document builder and XPath factories through `org.apache.commons:commons-secure-xml` 1.0.0 on the default path. The secure factories enable XML secure processing and install a non-removable entity-resolver floor: external DTD and entity lookups are resolved to empty content instead of being fetched, and internal entity expansion is bounded. The documentation for `XmlStringLookup.secure=false` system property is removed and was not implemented in git master.

The secure-path tests are adapted to the secure contract: a parser may either reject a document with an external reference or parse it with the reference resolved to empty content, so they now assert that the external content does not leak into the result instead of expecting one fixed failure mode. CI and CodeQL run with `-Puse-apache-snapshots` so the SNAPSHOT dependency resolves.

🤖 Generated with [Claude Code](https://claude.com/claude-code) + Gary's edits.

### CODE REVIEW

File: src/main/java/org/apache/commons/text/lookup/XmlStringLookup.java [hunk 1]
```diff
@@ -128,14 +130,21 @@ public String lookup(final String key) {
         }
         final String documentPath = keys[0];
         final String xpath = StringUtils.substringAfterLast(key, SPLIT_CH);
-        final DocumentBuilderFactory dbFactory = DocumentBuilderFactory.newInstance();
+        // The secure factory installs a non-removable resolver floor that ignores the JAXP access properties,
+        // so the documented opt-outs keep a plain factory: a feature map without secure processing, or the
+        // standard javax.xml.accessExternalDTD system property re-allowing external access.
+        final boolean secure = Boolean.TRUE.equals(xmlFactoryFeatures.get(XMLConstants.FEATURE_SECURE_PROCESSING))
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/text/lookup/XmlStringLookup.java]: @ppkarwasz 
Isn't this handled by the library now?

File: src/test/java/org/apache/commons/text/lookup/XmlStringLookupTest.java [hunk 2]
```diff
@@ -50,6 +51,19 @@ class XmlStringLookupTest {
...
 
+    /**
+     * Asserts the secure contract for an external reference: the parser either rejects the document or parses it
+     * with the reference resolved to empty content, but the external content never appears in the result.
+     */
+    static void assertBlocksOrDoesNotLeak(final Supplier<String> lookup, final String external) {
+        try {
+            final String result = lookup.get();
+            assertFalse(result != null && result.contains(external), () -> "external content leaked: " + result);
+        } catch (final IllegalArgumentException e) {
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/text/lookup/XmlStringLookupTest.java]: @ppkarwasz Is this expected? If so, use `assertThrows()`.

File: src/main/java/org/apache/commons/text/lookup/XmlStringLookup.java [hunk 3]
```diff
@@ -128,14 +130,21 @@ public String lookup(final String key) {
...
+                && StringUtils.isEmpty(System.getProperty("javax.xml.accessExternalDTD"));
+        final DocumentBuilderFactory dbFactory = secure ? SecureDocumentBuilderFactory.newInstance() : DocumentBuilderFactory.newInstance();
         try {
             for (final Entry<String, Boolean> p : xmlFactoryFeatures.entrySet()) {
                 dbFactory.setFeature(p.getKey(), p.getValue());
             }
             try (InputStream inputStream = Files.newInputStream(getPath(documentPath))) {
                 final Document doc = dbFactory.newDocumentBuilder().parse(inputStream);
-                final XPathFactory xpFactory = XPathFactory.newInstance();
+                final XPathFactory xpFactory = Boolean.TRUE.equals(xPathFactoryFeatures.get(XMLConstants.FEATURE_SECURE_PROCESSING))
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/text/lookup/XmlStringLookup.java]: Same question as above.

File: src/main/java/org/apache/commons/text/lookup/XmlStringLookup.java (hunk 1 above)
Comment (ppkarwasz) [reply src/main/java/org/apache/commons/text/lookup/XmlStringLookup.java]: This looks like an attempt by the agent to keep the `*ExternalEntityOn` tests working. Those tests rely on the specifics of the hardening to allow fetching external entities. On the built-in JDK parser:

- Setting FSP to `false`,
- or setting `javax.xml.accessExternalDTD` to "all",

is enough to reenable external entity fetching.

Commons Secure XML use a resolver to block the fetch instead, so the technique above will fail, since the resolver will block before `javax.xml.accessExternalDTD` is even evaluated.

I will of course clean up these (they don't make sense), but we might need a way to allow users to specify external fetches. My proposal is to use a resolver that uses the nested `PathFence` to determine if the fetch is allowed or not. WDYT?

Comment (garydgregory) [reply src/main/java/org/apache/commons/text/lookup/XmlStringLookup.java]: So the test should configure the objects for its own needs and leave the main code to factory class changes only, right?

Comment (ppkarwasz) [reply src/main/java/org/apache/commons/text/lookup/XmlStringLookup.java]: Yes, although right now there is not way to have a test with “external entities on”.

In one of the tests I see a property `XmlStringLookup.secure` being set, but I don't see that property being used anywhere in the code: was there every such a property?

Comment (garydgregory) [reply src/main/java/org/apache/commons/text/lookup/XmlStringLookup.java]: Interesting, I see it referred to from Javadoc comments at

https://github.com/apache/commons-text/blob/5d882dca7cb01b52a184a7e36f71d0aae4b1c9ac/src/main/java/org/apache/commons/text/lookup/StringLookupFactory.java#L1630 

and

https://github.com/apache/commons-text/blob/5d882dca7cb01b52a184a7e36f71d0aae4b1c9ac/src/main/java/org/apache/commons/text/lookup/StringLookupFactory.java#L1674

but _not_ from code, which tell me the comments must be left over from an earlier implementation or is was an incomplete implementation.

Comment (ppkarwasz) [reply src/main/java/org/apache/commons/text/lookup/XmlStringLookup.java]: Should we implement `XmlStringLookup.secure` then? Or a resolver that uses `PathFence`? Or maybe we should use both?

Comment (garydgregory) [reply src/main/java/org/apache/commons/text/lookup/XmlStringLookup.java]: Well, right now, you do get secure processing enabled by default in `master`, and unlike what the Javadoc says, it doesn't appear you can turn it off with a sys prop, BUT, you can set custom features (`StringLookupFactory.xmlStringLookup(Map<String, Boolean>, Path...)` which ends up calling `setFeature()`) so you can shoot yourself in the foot to your heart's desire.

The simple path would be:
1. Update the Javadoc to reflect the code: no sys prop, unless you can find it used ;)
2. Redo the main side of this PR to use Commons Secure XML cleanly and without leaving cruft behind

For example, I think the statics `DEFAULT_*` can go away

Path fences are mixing concerns for this PR. We want to integrate Secure XML and then see if we need improvements in Secure XML and/or call sites, later if possible.

Comment (ppkarwasz) [reply src/main/java/org/apache/commons/text/lookup/XmlStringLookup.java]: In https://github.com/apache/commons-text/pull/766/commits/bc9c20447a11dfbde21defdcc7713a62a0ba7ee8 I solved the problem by documenting that external entities can **not** be reenabled. Tests that relied on external entities where disabled, instead of being removed.

We'll need a follow-up PR, if we think users need to access external entities.

File: src/test/java/org/apache/commons/text/lookup/XmlStringLookupTest.java (hunk 2 above)
Comment (ppkarwasz) [reply src/test/java/org/apache/commons/text/lookup/XmlStringLookupTest.java]: Fixed in https://github.com/apache/commons-text/pull/766/commits/bc9c20447a11dfbde21defdcc7713a62a0ba7ee8

`assertBlocksOrDoesNotLeak` intentionally does not use `assertThrows()`: we really don't care whether parsing an XML document referencing an external entity throws. The important part is that the content is not fetched.

In the refactored version I replaced it with `asserDoesNotLeak`, so throwing is no longer an option.

File: src/main/java/org/apache/commons/text/lookup/XmlStringLookup.java (hunk 3 above)
Comment (ppkarwasz) [reply src/main/java/org/apache/commons/text/lookup/XmlStringLookup.java]: Fixed in https://github.com/apache/commons-text/pull/766/commits/bc9c20447a11dfbde21defdcc7713a62a0ba7ee8

File: src/test/java/org/apache/commons/text/lookup/StringLookupFactoryTest.java [hunk 4]
```diff
@@ -293,12 +293,13 @@ void testXmlStringLookup() {
...
-        assertThrows(IllegalArgumentException.class,
-                () -> StringLookupFactory.INSTANCE.xmlStringLookup().apply(XmlStringLookupTest.DOC_DIR + "document-entity-ref.xml:/document/content"));
+        XmlStringLookupTest.assertDoesNotLeak(
+                () -> StringLookupFactory.INSTANCE.xmlStringLookup().apply(XmlStringLookupTest.DOC_DIR + "document-entity-ref.xml:/document/content"),
+                XmlStringLookupTest.DATA);
     }
 
     @Test
-    @SetSystemProperty(key = "XmlStringLookup.secure", value = "false")
+    @Disabled("External entities are blocked by Commons Secure XML through an entity resolver and can no longer be re-enabled.")
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/text/lookup/StringLookupFactoryTest.java]: That's not right. There should be 2 tests instead of disabling this one. 1) this test is enabled and asserts the new default behavior. 2) shows how to override the behavior to get the old one back.

**Review by garydgregory — CHANGES_REQUESTED** [review]

File: src/test/java/org/apache/commons/text/lookup/XmlStringLookupTest.java [hunk 5]
```diff
@@ -64,25 +75,27 @@ void testBadXPath() {
...
     void testInterpolatorExternalDtdOff() {
         final StringSubstitutor stringSubstitutor = StringSubstitutor.createInterpolator();
-        assertThrows(IllegalArgumentException.class, () -> stringSubstitutor.replace("${xml:" + DOC_DIR + "document-external-dtd.xml:/document/content}"));
+        assertDoesNotLeak(() -> stringSubstitutor.replace("${xml:" + DOC_DIR + "document-external-dtd.xml:/document/content}"),
+                "This is an external entity.");
     }
 
     @Test
     @SetSystemProperty(key = "javax.xml.accessExternalDTD", value = "file")
+    @Disabled("External entities are blocked by Commons Secure XML through an entity resolver and can no longer be re-enabled.")
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/text/lookup/XmlStringLookupTest.java]: Same comment as above.

File: src/test/java/org/apache/commons/text/lookup/XmlStringLookupTest.java [hunk 6]
```diff
@@ -91,11 +104,12 @@ void testInterpolatorExternalDtdOn() {
     @Test
     void testInterpolatorExternalEntityOff() {
         final StringSubstitutor stringSubstitutor = StringSubstitutor.createInterpolator();
-        assertThrows(IllegalArgumentException.class, () -> stringSubstitutor.replace("${xml:" + DOC_DIR + "document-entity-ref.xml:/document/content}"));
+        assertDoesNotLeak(() -> stringSubstitutor.replace("${xml:" + DOC_DIR + "document-entity-ref.xml:/document/content}"), DATA);
     }
 
     @Test
     @SetSystemProperty(key = "javax.xml.accessExternalDTD", value = "file")
+    @Disabled("External entities are blocked by Commons Secure XML through an entity resolver and can no longer be re-enabled.")
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/text/lookup/XmlStringLookupTest.java]: Same comment as above.

File: src/test/java/org/apache/commons/text/lookup/StringLookupFactoryTest.java (hunk 4 above)
Comment (ppkarwasz) [reply src/test/java/org/apache/commons/text/lookup/StringLookupFactoryTest.java]: There is no way to restore the previous behavior: the public API allows setting arbitrary **features**, but does not offer a way to call `setEntityResolver`, which is the only way to opt-in some external resources.

I thought we agreed in https://github.com/apache/commons-text/pull/766#discussion_r3943978584 to introduce a mechanism that allows external fetches in a follow-up PR, but I am happy to do it in this PR.

What mechanism do you think is appropriate?

- A system property to use `DocumentBuilderFactory.newInstance()` instead of Commons Secure XML?
- Extend the `factoryFeatures` argument of [`xmlStringLookup`](https://javadoc.io/doc/org.apache.commons/commons-text/latest/org/apache/commons/text/lookup/StringLookupFactory.html) with a special key that enables external entities?
- Restrict external entities to the same folders as the XML file itself (i.e. path fences)?

Comment (garydgregory) [reply src/test/java/org/apache/commons/text/lookup/StringLookupFactoryTest.java]: Hi @ppkarwasz 
I think you mean [this specific XML method](https://javadoc.io/doc/org.apache.commons/commons-text/latest/org/apache/commons/text/lookup/StringLookupFactory.html#xmlStringLookup(java.util.Map,java.nio.file.Path...)) in `StringLookupFactory` which already accepts a `Path[]` for fences, so that would map nicely for the resolver to only accept external entities in those paths. WDYT?

Otherwise, we could introduce a system property to bypass Commons Secure XML which would still let some people complain that we are insecure in some cases. Well, we would reintroduce it since we had it in the Javadoc before but it was never implemented.

So I am favoring reusing the path fence to also fence external entities.

Comment (ppkarwasz) [reply src/test/java/org/apache/commons/text/lookup/StringLookupFactoryTest.java]: Cool, I have a prototype of the `PathFenceResolver` locally.

There is only one edge case to solve: if the `Path[]` is `null`, does it mean entities can be in any `file:` or in no file?

Passing `null` as `Path[]` means the XML file can be anywhere on the filesystem, but that is probably fine (`/etc/passwd` is not an XML file). On the other hand external entities don't have a format, so it might be more prudent disallow entities unless a non-empty fence has been created.

Comment (garydgregory) [reply src/test/java/org/apache/commons/text/lookup/StringLookupFactoryTest.java]: Currently, null is mapped to an empty array in `org.apache.commons.text.lookup.PathFence.Builder.setRoots(Path...)` which means that anything tested is rejected. So it behaves like a grant-based system, which is good I think.

Comment (ppkarwasz) [reply src/test/java/org/apache/commons/text/lookup/StringLookupFactoryTest.java]: That is not what I read from the code:

https://github.com/apache/commons-text/blob/9c9a0047b2c4ae5b6be0382177cd6d19439601f7/src/main/java/org/apache/commons/text/lookup/PathFence.java#L96-L99

So I would propose:

1. If `Path[]` is empty, then the XML can be fetched, but no external entities.
2. If `Path[]` is not empty, then both the XML and external entities must be inside the fence.

Comment (garydgregory) [reply src/test/java/org/apache/commons/text/lookup/StringLookupFactoryTest.java]: You are correct! :) I just added some tests for the public API to this PR and stepped through that part of the code in a debugger.

I agree with your 2 points :) Progress!

Comment (ppkarwasz) [reply src/test/java/org/apache/commons/text/lookup/StringLookupFactoryTest.java]: Implemented in https://github.com/apache/commons-text/pull/766/commits/3014afc4b625a44e845d10e6dd154ec37ae830a5

A **fenced** lookup resolves external resources from the fences, an **unfenced** one ignore them.

File: src/test/java/org/apache/commons/text/lookup/XmlStringLookupTest.java (hunk 5 above)
Comment (ppkarwasz) [reply src/test/java/org/apache/commons/text/lookup/XmlStringLookupTest.java]: From the comment I can not see the test name any more, but `javax.xml.accessExternalDTD`-based tests were removed in https://github.com/apache/commons-text/pull/766/commits/3014afc4b625a44e845d10e6dd154ec37ae830a5.

There were three kinds of tests:

1. An external resource (e.g. DTD) is blocked by default.
2. You can override that with `javax.xml.accessExternalDTD`.
3. An external resource is allowed if FSP is turned off.

Now the second test case is impossible (no Java system property override), while the third kind doesn't use FSP, but a fence.

File: src/test/java/org/apache/commons/text/lookup/XmlStringLookupTest.java (hunk 6 above)
Comment (ppkarwasz) [reply src/test/java/org/apache/commons/text/lookup/XmlStringLookupTest.java]: Same as https://github.com/apache/commons-text/pull/766#discussion_r3965880979

**Comment by garydgregory** [comment]
Merged 🚀 Thank you @ppkarwasz
